File: FakturaCZ-Windows/app/models/subject.py
"""Subject (client/supplier) data model."""

from __future__ import annotations

import json
from dataclasses import dataclass, field, asdict
from datetime import datetime
from typing import Optional


@dataclass
class Subject:
    """A business subject identified by Czech registration number (IČO)."""

    # Identifiers
    registration_no: str  # IČO — 8-digit Czech business ID

    # Business details (enriched from ARES)
    name: str = ""
    street: str = ""
    city: str = ""
    zip: str = ""
    vat_no: Optional[str] = None  # DIČ

    # Contact (user-provided)
    email: Optional[str] = None
    phone: Optional[str] = None
    bank_account: Optional[str] = None
    iban: Optional[str] = None

    # Metadata
    created_at: str = field(default_factory=lambda: datetime.now().isoformat())
    updated_at: str = field(default_factory=lambda: datetime.now().isoformat())
    last_ares_sync: Optional[str] = None

    @property
    def is_vat_payer(self) -> bool:
        return bool(self.vat_no)

    @property
    def formatted_address(self) -> str:
        parts = [self.street, f"{self.zip} {self.city}"]
        return "\n".join(p for p in parts if p.strip())

    def apply_ares_data(self, data: dict) -> None:
        """Apply data retrieved from the ARES registry."""
        self.name = data.get("name", self.name)
        self.street = data.get("street", self.street)
        self.city = data.get("city", self.city)
        self.zip = data.get("zip", self.zip)
        self.vat_no = data.get("vat_no")
        self.last_ares_sync = datetime.now().isoformat()
        self.updated_at = datetime.now().isoformat()

    def to_dict(self) -> dict:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict) -> Subject:
        return cls(**{k: v for k, v in data.items() if k in cls.__dataclass_fields__})

File: FakturaCZ/FakturaCZ/Models/Subject.swift
import Foundation
import SwiftData

/// A business subject (client/supplier) identified by Czech registration number (IČO).
@Model
final class Subject {
    // MARK: - Identifiers

    /// Czech registration number (IČO) — the primary business identifier.
    @Attribute(.unique)
    var registrationNo: String

    // MARK: - Business Details (enriched from ARES)

    var name: String
    var street: String
    var city: String
    var zip: String

    /// Czech VAT number (DIČ), nil if not a VAT payer.
    var vatNo: String?

    /// Derived flag: true if vatNo is non-nil and non-empty.
    var isVATPayer: Bool {
        guard let vat = vatNo else { return false }
        return !vat.isEmpty
    }

    // MARK: - Contact (user-provided)

    var email: String?
    var phone: String?
    var bankAccount: String?
    var iban: String?

    // MARK: - Metadata

    var createdAt: Date
    var updatedAt: Date

    /// Date of last successful ARES sync.
    var lastARESSync: Date?

    // MARK: - Relationships

    @Relationship(deleteRule: .deny, inverse: \Invoice.subject)
    var invoices: [Invoice] = []

    // MARK: - Init

    init(
        registrationNo: String,
        name: String = "",
        street: String = "",
        city: String = "",
        zip: String = "",
        vatNo: String? = nil,
        email: String? = nil,
        phone: String? = nil,
        bankAccount: String? = nil,
        iban: String? = nil
    ) {
        self.registrationNo = registrationNo
        self.name = name
        self.street = street
        self.city = city
        self.zip = zip
        self.vatNo = vatNo
        self.email = email
        self.phone = phone
        self.bankAccount = bankAccount
        self.iban = iban
        self.createdAt = Date()
        self.updatedAt = Date()
    }

    // MARK: - ARES Update

    /// Apply data retrieved from the ARES registry.
    func applyARESData(_ data: ARESData) {
        self.name = data.name
        self.street = data.street
        self.city = data.city
        self.zip = data.zip
        self.vatNo = data.vatNo
        self.lastARESSync = Date()
        self.updatedAt = Date()
    }

    // MARK: - Display

    var formattedAddress: String {
        [street, "\(zip) \(city)"].filter { !$0.isEmpty }.joined(separator: "\n")
    }
}
